# 遞迴數列（高中數學第二冊 1-2）

### 課本例題： 階層
數列 $\{a_n\}$ 的遞迴定義是 $\left\{\begin{array}{l}a_1=1,\\a_n=na_{n-1},n\in \mathbb{N} ,n\ge 2\end{array}\right.$，求 $a_5$

In [1]:
def myfac(x):
    if x==1:
        return 1            #當遇到 1! 時，就會回傳 1
    else :
        return x*myfac(x-1) #回傳值是 x * (x-1)! ， 程式會繼續求 (x-1)!


In [2]:
myfac(5)

120

以 myfac(5) 來說 ： <br>
myfac(5) 會回傳 5 x myfac(4) <br>
myfac(4) 會回傳 4 x myfac(3) <br>
myfac(3) 會回傳 3 x myfac(2) <br>
myfac(2) 會回傳 2 x myfac(1) <br>
myfac(1) 會回傳 1 <br>
依此回推就會得到 $5\times 4 \times 3 \times 2\times 1 =120$


注意：這個寫法的定義域必須為 $\mathbb{N}$ ，如果硬要算 0! 或是 (-1)! 都會出錯。可以用 ``if`` 來確保輸入的定義域是否正確。

### 隨堂練習： 
數列 $\{a_n\}$ 的遞迴定義是 $\left\{\begin{array}{l}a_1=2,\\a_n=(n-1)a_{n-1},n\in \mathbb{N} ,n\ge 2\end{array}\right.$，求 $a_8$

In [3]:
def pract1(x):
    if x==1:
        return 2
    else :
        return (x-1)*pract1(x-1)


In [4]:
pract1(8)

10080

### 隨堂練習：費氏數列 (Fibonacci sequence)
數列 $\{a_n\}$ 的遞迴定義是 $\left\{\begin{array}{l}a_1=1,\\a_2=1\\a_n=a_{n-1}+a_{n-2},n\in \mathbb{N} ,n\ge 3\end{array}\right.$，求 $a_{20}$


Hint: <BR>
``if x==1:
elif x==2:
else :``

In [5]:
def fibonacci(x):
    if x==1:
        return 1
    elif x==2:
        return 1
    else :
        return fibonacci(x-1)+fibonacci(x-2)



In [6]:
fibonacci(20)

6765